evap/staff/tools.py

```diff
@@ -68,31 +69,95 @@ def delete_navbar_cache_for_users(users):
         cache.delete(key)
 
 
-def bulk_delete_users(request, username_file, test_run):
-    usernames = [u.decode().strip() for u in username_file.readlines()]
-    users = UserProfile.objects.exclude(username__in=usernames)
-    deletable_users = [u for u in users if u.can_be_deleted_by_manager]
-    users_to_mark_inactive = [u for u in users if u.is_active and not u.can_be_deleted_by_manager and u.can_be_marked_inactive_by_manager]
-
-    messages.info(request, _('The uploaded text file contains {} usernames. {} other users have been found in the database. '
-                           'Of those, {} will be deleted and {} will be marked inactive.')
-                  .format(len(usernames), len(users), len(deletable_users), len(users_to_mark_inactive)))
-    messages.info(request, mark_safe(_('Users to be deleted are:<br />{}')
-                  .format('<br />'.join([u.username for u in deletable_users]))))
-    messages.info(request, mark_safe(_('Users to be marked inactive are:<br />{}')
-                  .format('<br />'.join([u.username for u in users_to_mark_inactive]))))
+def find_matching_internal_user_for_email(request, email):
+    # for internal users only the part before the @ must be the same to match a user to an email
+    matching_users = [
+        user for user
+        in UserProfile.objects.filter(email__startswith=email.split('@')[0] + '@')
+        if not user.is_external
+    ]
+
+    if not matching_users:
+        return None
+    if len(matching_users) > 1:
+        messages.warning(request,
+            _('For {} multiple matching existing users have been found. {} will be updated with the new data.')
+            .format(email, matching_users[0].email)
+        )
+    return matching_users[0]
+
+
+def bulk_update_users(request, user_file, test_run):
+    # pylint: disable=too-many-branches
+    # user_file must have one user per line in the format "{username},{email}"
+    imported_emails = {clean_email(line.decode().split(',')[1]) for line in user_file.readlines()}
+
+    emails_of_users_to_be_created = []
+    users_to_be_updated = []
+    emails_of_non_obsolete_users = list(imported_emails)
+    skipped_external_emails_counter = 0
+
+    for imported_email in imported_emails:
+        if is_external_email(imported_email):
+            skipped_external_emails_counter += 1
+            continue
+        matching_user = find_matching_internal_user_for_email(request, imported_email)
+        if not matching_user:
+            emails_of_users_to_be_created.append(imported_email)
+        elif matching_user.email != imported_email:
+            users_to_be_updated.append((matching_user, imported_email))
+            emails_of_non_obsolete_users.append(matching_user.email)  # add old address so this user won't be deleted
```

Comment:
Could you move the `emails_of_non_obsolete_users` variable just before `obsolete_users`?
It could be expressed with something similar to:
```python
emails_of_non_obsolete_users = set(imported_emails) | {user.email for user, _imported_email in users_to_be_updated}
```